embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,887 @@
+    /// possible for the corresponding receiver to hang up immediately after
+    /// this function returns `Ok`.
+    ///
+    /// # Errors
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`close`]: Receiver::close
+    /// [`Receiver`]: Receiver
+    pub async fn send(&self, message: T) -> Result<(), SendError<T>> {
+        SendFuture {
+            sender: self.clone(),
+            message: UnsafeCell::new(message),
+        }
+        .await
+    }
+
+    /// Attempts to immediately send a message on this `Sender`
+    ///
+    /// This method differs from [`send`] by returning immediately if the channel's
+    /// buffer is full or no receiver is waiting to acquire some data. Compared
+    /// with [`send`], this function has two failure cases instead of one (one for
+    /// disconnection, one for a full buffer).
+    ///
+    /// # Errors
+    ///
+    /// If the channel capacity has been reached, i.e., the channel has `n`
+    /// buffered values where `n` is the argument passed to [`channel`], then an
+    /// error is returned.
+    ///
+    /// If the receive half of the channel is closed, either due to [`close`]
+    /// being called or the [`Receiver`] handle dropping, the function returns
+    /// an error. The error includes the value passed to `send`.
+    ///
+    /// [`send`]: Sender::send
+    /// [`channel`]: channel
+    /// [`close`]: Receiver::close
+    pub fn try_send(&self, message: T) -> Result<(), TrySendError<T>> {
+        self.channel.get().try_send(message)
+    }
+
+    /// Completes when the receiver has dropped.
+    ///
+    /// This allows the producers to get notified when interest in the produced
+    /// values is canceled and immediately stop doing work.
+    pub async fn closed(&self) {
+        CloseFuture {
+            sender: self.clone(),
+        }
+        .await
+    }
+
+    /// Checks if the channel has been closed. This happens when the
+    /// [`Receiver`] is dropped, or when the [`Receiver::close`] method is
+    /// called.
+    ///
+    /// [`Receiver`]: crate::sync::mpsc::Receiver
+    /// [`Receiver::close`]: crate::sync::mpsc::Receiver::close
+    pub fn is_closed(&self) -> bool {
+        self.channel.get().is_closed()
+    }
+}
+
+struct SendFuture<'ch, M, T, const N: usize>
+where
+    M: Mutex<Data = ()>,
+{
+    sender: Sender<'ch, M, T, N>,
+    message: UnsafeCell<T>,
+}
+
+impl<'ch, M, T, const N: usize> Future for SendFuture<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = Result<(), SendError<T>>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self
+            .sender
+            .channel
+            .get()
+            .try_send_with_context(unsafe { self.message.get().read() }, Some(cx))
```

Comment:
It would be simpler to have `poll_send` in `Channel` and just call it, so you can avoid converting the error here, and the issues with moving the T around.

Reply:
I no longer have issues copying T around given commit ae572e94307d226b3ea45e878724694d2528affa. Do you think this still applies? I've got some nice re-use going with `try_send_with_context`.

Reply:
Yup, looks correct now!

Reply:
Making everything a single function might help in moving it less, but we can optimize that later. Let's get a working, correct version merged first! Marking as resolved